cache/network/api.go

```diff
@@ -0,0 +1,230 @@
+package network
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+const maxKeyLength = 512
+const maxKeyCount = 8
+
+type prepareUploadRequest struct {
+	CacheKey           string `json:"cache_key"`
+	ArchiveFileName    string `json:"archive_filename"`
+	ArchiveContentType string `json:"archive_content_type"`
+	ArchiveSizeInBytes int64  `json:"archive_size_in_bytes"`
+}
+
+type prepareUploadResponse struct {
+	ID            string            `json:"id"`
+	UploadMethod  string            `json:"method"`
+	UploadURL     string            `json:"url"`
+	UploadHeaders map[string]string `json:"headers"`
+}
+
+type restoreResponse struct {
+	URL string `json:"url"`
+}
+
+type apiClient struct {
+	httpClient  *retryablehttp.Client
+	baseURL     string
+	accessToken string
+}
+
+func newAPIClient(client *retryablehttp.Client, baseURL string, accessToken string) apiClient {
+	return apiClient{
+		httpClient:  client,
+		baseURL:     baseURL,
+		accessToken: accessToken,
+	}
+}
+
+func (c apiClient) prepareUpload(requestBody prepareUploadRequest) (prepareUploadResponse, error) {
+	url := fmt.Sprintf("%s/upload", c.baseURL)
+
+	body, err := json.Marshal(requestBody)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+
+	req, err := retryablehttp.NewRequest("POST", url, body)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+	req.Header.Set("Content-type", "application/json")
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	if resp.StatusCode != http.StatusCreated {
+		errorResp, err := ioutil.ReadAll(resp.Body)
```

Comment:
Is the response body a raw message in the case of an error and not a structured message?

Reply:
It is a structured JSON, but I believe it's fine to print it as raw text when there is an error.